processing/src/main/java/io/druid/segment/DimensionHandlerUtils.java

```diff
@@ -138,15 +141,21 @@ public static DimensionHandler getHandlerFromCapabilities(
   // (e.g., treating missing columns as empty String columns)
   private static ColumnCapabilities getEffectiveCapabilities(
       String dimName,
-      ColumnCapabilities capabilities
+      ColumnCapabilities capabilities,
+      boolean hasExFn
   )
   {
     if (capabilities == null) {
       capabilities = DEFAULT_STRING_CAPABILITIES;
     }
 
-    // non-Strings aren't actually supported yet
-    if (capabilities.getType() != ValueType.STRING) {
+    // Complex dimension type is not supported
+    if (capabilities.getType() == ValueType.COMPLEX) {
+      capabilities = DEFAULT_STRING_CAPABILITIES;
+    }
+
+    // Currently, all extractionFns output Strings, so force the type if present
```

Comment:
Changed comment to:

"Currently, all extractionFns output Strings, so the column will return String values via a DimensionSelector if an extractionFn is present."